Add ActiveStash getting started to readme (#34)
cipherstash/activestash: README.md

# ActiveStash

[ActiveStash](https://cipherstash.com/activestash) is the Rails specific gem for using [CipherStash](https://cipherstash.com).

> It provides search functionality for ActiveRecord models that are configured to use field level encryption (using [Lockbox](https://github.com/ankane/lockbox) or
> [EncryptedRecord](https://guides.rubyonrails.org/active_record_encryption.html)).

Reviewer: ```suggestion
ActiveStash gives you encryption search on ActiveRecord models using application level encryption (using libraries like [Lockbox](https://github.com/ankane/lockbox) and [ActiveRecord Encryption](https://guides.rubyonrails.org/active_record_encryption.html)).
```

# ActiveStash

[ActiveStash](https://cipherstash.com/activestash) is the Rails-specific gem for using [CipherStash](https://cipherstash.com).

ActiveStash gives you encryption search on ActiveRecord models using application level encryption (using libraries like [Lockbox](https://github.com/ankane/lockbox) and [ActiveRecord Encryption](https://guides.rubyonrails.org/active_record_encryption.html)).

> When records are created or updated, they are indexed into a CipherStash collection
> which can be queried via an ActiveStash::Relation.

Reviewer: ```suggestion
When records are created or updated, they are indexed into a CipherStash collection which can be queried via an `ActiveStash::Relation`.
```

Reviewer: Tiny style thing: we make sure we have one newline per sentence. 

Makes it easier for future reviewers to parse mid-sentence changes.

2. Create a new rails app, the example app is called `lockboxActiveStash`.

```bash
rails new lockboxActiveStash
```

3. Add a user scaffold

```bash
rails generate scaffold user name email:uniq dob:date
```

4. Setup lockbox

* Add the lockbox dependancy to your gemfile

```
gem "lockbox"
```

* Run bundle install

```bash
➜  bundle install
```

* Generate a key from lockbox via rails console

```bash
➜  rails c
Loading development environment (Rails 7.0.3)
irb(main):001:0> Lockbox.generate_key
=> "<copy the key that is generated here>"
```

* Copy the key that lockbox generated and exit out of rails console.

*  Save the key generated by lockbox as a rails credential

```bash
# EDITOR='<enter name of editor here e.g vim | code | nano> --wait' rails credentials:edit
EDITOR='code --wait' rails credentials:edit
```

This will open the credentials file with the editor specified in the command.

* Add the lockbox key to the file in the format below.

```
lockbox:
    master_key: "Paste the key generated from lockbox in here"
```

* Save the file and close. You should see the below confirmation in your terminal.

```bash

> File encrypted and saved.
> ```
> 
> 5. Encrypt your user model using lockbox
> 
> * Generate a migration to add the ciphertext fields to the user model.
> 
>   This is where the data encrypted by lockbox is saved.
> 
> ```bash
> ➜  rails generate migration AddLockboxCiphertextToUser email_ciphertext:string name_ciphertext:string dob_ciphertext:string
> ➜  rails db:migrate
> ```
> 
> * Update your user model
> 
> Add the fields that are encrypted to the user model.
> 
> 
> ```ruby
> # app/models/user.rb
>   class User < ApplicationRecord
>     lockbox_encrypts :name, :email
>     lockbox_encrypts :dob, type: :date
> 
>     self.ignored_columns = %w[name email dob]
>   end
> ```
> 
> Note: With Rails 6 and below, use encrypts instead of lockbox_encrypts (Rails 7)
> 
> We now have lockbox setup, when a user is created the data is encrypted and stored in the ciphertext fields we added in the previous step.
> 
> ```ruby
> $ rails console
>  irb(main):001:0> User.create(name: "Grace Hopper", email: "grace@example.com", dob: Date.new(1906,12,9))
>    (0.5ms)  SELECT sqlite_version(*)
>   TRANSACTION (0.0ms)  begin transaction                                                                                  
>   User Create (0.6ms)  INSERT INTO "users" ("created_at", "updated_at", "email_ciphertext", "name_ciphertext", "dob_ciphertext") VALUES (?, ?, ?, ?, ?)  [["created_at", "2022-05-26 05:29:31.894710"], ["updated_at", "2022-05-26 05:29:31.894710"], ["email_ciphertext", "LtsjpkXfGMlpOtAQyqUBeRy2PcNPNk8LJIR5JAli+uwp9/QkN+m0MUXRujOf"], ["name_ciphertext", "cmrgiaJ0UFbj9a1+nyr7jLJe/ATc5NCIO0k7a2T4uf2w1urgn13aqg=="], ["dob_ciphertext", "EYREKL1kJR6pjwWvLPvCcKujHb39cWQLkmsrc+//ETG9FGyrdXM="]]                                                                                 
>   TRANSACTION (0.5ms)  commit transaction                                                                                 
> => #<User id: 1, created_at: "2022-05-26 05:29:31.894710000 +0000", updated_at: "2022-05-26 05:29:31.894710000 +0000", email: [FILTERED], name: [FILTERED], dob: [FILTERED]>      
> ```
> 
> Although our stored data is secure, we are not able to query it.
> 
> ```bash
> irb(main):003:0> User.where(email: "grace@hopper.example").count
>   User Count (0.4ms)  SELECT COUNT(*) FROM "users" WHERE "users"."email" = ?  [["email", "grace@hopper.example"]]
> => 0
> ```
> 
> This is what ActiveStash solves.
> 
> Every time you create or update a record, the data is indexed to CipherStash via ActiveRecord callbacks.
> 
> Queries are delegated to CipherStash but return ActiveRecord models so things just work.
> 
> Go to [getting started with ActiveStash](#getting-started-with-activestash) to learn how.
> 
> 

Reviewer: ```suggestion
```

> File encrypted and saved.
> ```
> 
> 5. Encrypt your user model using lockbox
> 
> * Generate a migration to add the ciphertext fields to the user model.
> 
>   This is where the data encrypted by lockbox is saved.
> 
> ```bash
> ➜  rails generate migration AddLockboxCiphertextToUser email_ciphertext:string name_ciphertext:string dob_ciphertext:string
> ➜  rails db:migrate
> ```
> 
> * Update your user model
> 
> Add the fields that are encrypted to the user model.
> 
> 
> ```ruby
> # app/models/user.rb
>   class User < ApplicationRecord
>     lockbox_encrypts :name, :email
>     lockbox_encrypts :dob, type: :date
> 
>     self.ignored_columns = %w[name email dob]
>   end
> ```
> 
> Note: With Rails 6 and below, use encrypts instead of lockbox_encrypts (Rails 7)
> 
> We now have lockbox setup, when a user is created the data is encrypted and stored in the ciphertext fields we added in the previous step.
> 
> ```ruby
> $ rails console
>  irb(main):001:0> User.create(name: "Grace Hopper", email: "grace@example.com", dob: Date.new(1906,12,9))
>    (0.5ms)  SELECT sqlite_version(*)
>   TRANSACTION (0.0ms)  begin transaction                                                                                  
>   User Create (0.6ms)  INSERT INTO "users" ("created_at", "updated_at", "email_ciphertext", "name_ciphertext", "dob_ciphertext") VALUES (?, ?, ?, ?, ?)  [["created_at", "2022-05-26 05:29:31.894710"], ["updated_at", "2022-05-26 05:29:31.894710"], ["email_ciphertext", "LtsjpkXfGMlpOtAQyqUBeRy2PcNPNk8LJIR5JAli+uwp9/QkN+m0MUXRujOf"], ["name_ciphertext", "cmrgiaJ0UFbj9a1+nyr7jLJe/ATc5NCIO0k7a2T4uf2w1urgn13aqg=="], ["dob_ciphertext", "EYREKL1kJR6pjwWvLPvCcKujHb39cWQLkmsrc+//ETG9FGyrdXM="]]                                                                                 
>   TRANSACTION (0.5ms)  commit transaction                                                                                 
> => #<User id: 1, created_at: "2022-05-26 05:29:31.894710000 +0000", updated_at: "2022-05-26 05:29:31.894710000 +0000", email: [FILTERED], name: [FILTERED], dob: [FILTERED]>      
> ```
> 
> Although our stored data is secure, we are not able to query it.
> 
> ```bash
> irb(main):003:0> User.where(email: "grace@hopper.example").count
>   User Count (0.4ms)  SELECT COUNT(*) FROM "users" WHERE "users"."email" = ?  [["email", "grace@hopper.example"]]
> => 0
> ```
> 
> This is what ActiveStash solves.
> 
> Every time you create or update a record, the data is indexed to CipherStash via ActiveRecord callbacks.
> 
> Queries are delegated to CipherStash but return ActiveRecord models so things just work.
> 
> Go to [getting started with ActiveStash](#getting-started-with-activestash) to learn how.
> 
> 

Reviewer: ```suggestion

```

> File encrypted and saved.
> ```
> 
> 5. Encrypt your user model using lockbox
> 
> * Generate a migration to add the ciphertext fields to the user model.
> 
>   This is where the data encrypted by lockbox is saved.
> 
> ```bash
> ➜  rails generate migration AddLockboxCiphertextToUser email_ciphertext:string name_ciphertext:string dob_ciphertext:string
> ➜  rails db:migrate
> ```
> 
> * Update your user model
> 
> Add the fields that are encrypted to the user model.
> 
> 
> ```ruby
> # app/models/user.rb
>   class User < ApplicationRecord
>     lockbox_encrypts :name, :email
>     lockbox_encrypts :dob, type: :date
> 
>     self.ignored_columns = %w[name email dob]
>   end
> ```
> 
> Note: With Rails 6 and below, use encrypts instead of lockbox_encrypts (Rails 7)
> 
> We now have lockbox setup, when a user is created the data is encrypted and stored in the ciphertext fields we added in the previous step.
> 
> ```ruby
> $ rails console
>  irb(main):001:0> User.create(name: "Grace Hopper", email: "grace@example.com", dob: Date.new(1906,12,9))
>    (0.5ms)  SELECT sqlite_version(*)
>   TRANSACTION (0.0ms)  begin transaction                                                                                  
>   User Create (0.6ms)  INSERT INTO "users" ("created_at", "updated_at", "email_ciphertext", "name_ciphertext", "dob_ciphertext") VALUES (?, ?, ?, ?, ?)  [["created_at", "2022-05-26 05:29:31.894710"], ["updated_at", "2022-05-26 05:29:31.894710"], ["email_ciphertext", "LtsjpkXfGMlpOtAQyqUBeRy2PcNPNk8LJIR5JAli+uwp9/QkN+m0MUXRujOf"], ["name_ciphertext", "cmrgiaJ0UFbj9a1+nyr7jLJe/ATc5NCIO0k7a2T4uf2w1urgn13aqg=="], ["dob_ciphertext", "EYREKL1kJR6pjwWvLPvCcKujHb39cWQLkmsrc+//ETG9FGyrdXM="]]                                                                                 
>   TRANSACTION (0.5ms)  commit transaction                                                                                 
> => #<User id: 1, created_at: "2022-05-26 05:29:31.894710000 +0000", updated_at: "2022-05-26 05:29:31.894710000 +0000", email: [FILTERED], name: [FILTERED], dob: [FILTERED]>      
> ```
> 
> Although our stored data is secure, we are not able to query it.
> 
> ```bash
> irb(main):003:0> User.where(email: "grace@hopper.example").count
>   User Count (0.4ms)  SELECT COUNT(*) FROM "users" WHERE "users"."email" = ?  [["email", "grace@hopper.example"]]
> => 0
> ```
> 
> This is what ActiveStash solves.
> 
> Every time you create or update a record, the data is indexed to CipherStash via ActiveRecord callbacks.
> 
> Queries are delegated to CipherStash but return ActiveRecord models so things just work.
> 
> Go to [getting started with ActiveStash](#getting-started-with-activestash) to learn how.
> 
> 

Reviewer: ```suggestion

```

> ## Getting started with ActiveStash

Reviewer: ```suggestion
```

This will replace this heading with "Getting started" from line 40.

https://docs.cipherstash.com/tutorials/getting-started/create-an-account.html

```bash
➜  rake active_stash:login[YOURWORKSPACEID]
```

Note: If you are using `zsh` you may need to escape the brackets

```bash
rake active_stash:login\['WorkspaceId'\] 
```

4. Any model in which you include ActiveStash::Search, will need to have a `stash_id` column added of type `string`.

To add this to the table underlying your `User` model add the below migration:

```sh
$ rails g migration AddStashIdToUser stash_id:string:index
$ rails db:migrate
```

The above command also ensures that an index is created on `stash_id`.

5. Update your user model to include the ActiveStash search module and declare which fields are searchable.

```ruby
# app/models/user.rb
class User < ApplicationRecord
  include ActiveStash::Search

  encrypts :name, :email
  encrypts :dob, type: :date

  # Fields that will be indexed into CipherStash
  stash_index :name, :email, :dob

  self.ignored_columns = %w[name email dob]
end
```

6. Reindex your existing data into CipherStash with ActiveStash

```
➜  rails active_stash:reindexall
```

> 7. Try querying the same user we created earlier

Reviewer: ```suggestion
7. Query a user record:
```

```bash

> $ rails c
>  >> User.query(email: "grace@example.com").count
>  => 1

Reviewer: ```suggestion
$ rails c 
 >> User.where(email: "grace@example.com").count
 => 0 # no records, because the database isn't searchable
 >> User.query(email: "grace@example.com").count
 => 1 # a record, because CipherStash makes your encrypted database searchable$ rails c
```
